### PR TITLE
Add an INDEX to the Instance trait

### PR DESCRIPTION
**Pull request by shawntabrizi**
This PR introduces a simple `const INDEX: u8;` to the Instance trait used for Pallets.

This is useful for getting a minimal unique identifier for different instances, for example using it as a seed to some unique account generation.

The `DefaultInstance` uses `index == 0`, so we deprecate `Instance0` (which wasn't used anywhere) and instead use instances `1 ..= 16`.

### CODE REVIEW

File: frame/support/procedural/src/storage/instance_trait.rs [hunk 1]
```diff
@@ -83,6 +87,7 @@ pub fn decl_and_impl(scrate: &TokenStream, def: &DeclStorageDefExt) -> TokenStre
 				/// instance.
 				#[doc(hidden)]
 			),
+			index: 0,
```
Comment (shawntabrizi) [inline frame/support/procedural/src/storage/instance_trait.rs:90]: What is "inherent_instance"???

I gave it index 0 too, but technically that conflicts with `DefaultInstance`...?

Maybe it should be 255?

Comment (bkchr) [reply frame/support/procedural/src/storage/instance_trait.rs:90]: I would need to check the code, but somewhere we require an instance, even if we don't use any instancing. I think somewhere related to the genesis config.

Comment (kianenigma) [reply frame/support/procedural/src/storage/instance_trait.rs:90]: Based on the doc I would change this to 255. I am not sure when we need this instance though.

**Review by kianenigma — APPROVED** [review]

File: frame/support/procedural/src/storage/instance_trait.rs (hunk 1 above)
Comment (bkchr) [reply frame/support/procedural/src/storage/instance_trait.rs:90]: Yeah, I'm right. 

This is only used to make the type system happy, so the actual index isn't that important.

**Review by bkchr — APPROVED** [review]

File: frame/support/procedural/src/storage/instance_trait.rs (hunk 1 above)
Comment (shawntabrizi) [inline frame/support/procedural/src/storage/instance_trait.rs:90]:
```suggestion
			// This is just to make the type system happy. Not actually used.
			index: 0,
```

**Comment by shawntabrizi** [comment]
bot merge

**Comment by thiolliere** [comment]
note that pallet macro doesn't make use of instance trait anymore, instance trait was created in order to give different prefix to each instances, now that prefix is given by PalletInfo, the instance trait had no purpose so it was removed.

In order to have a INDEX identifier in pallet defines with pallet macro I think ppl should add it to the Config trait or we can add one in PalletInfo. (actually adding one in PalletInfo seems the best strategy to me).

**Comment by kianenigma** [comment]
FYI this is moving to the next polkadot release very soon, didn't we decide to revert this? cc @s3krit

**Comment by kianenigma** [comment]
(It doesn't have a big impact and we can remove it any point anyhow, but I just saw it in the release notes and thought better do it soon than later.)

**Comment by thiolliere** [comment]
The `Instance` trait will be deprecated alongside decl_storage macro.
It is not used in the pallet macro.

So for me it doesn't matter, we are not creating code debt here.
But if a revert PR opens I'll also approve.

**Comment by kianenigma** [comment]
> will be deprecated alongside decl_storage macro

Then I am also neutral, whatever @shawntabrizi prefers.

**Comment by shawntabrizi** [comment]
i dont think it matters if we keep it.

**Pull request merged**